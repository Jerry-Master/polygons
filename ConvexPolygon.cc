#include "ConvexPolygon.h"
#include <iostream>
// std::cout << "a1: " << this_angle << ", " << "a2: " << last_angle << std::endl;

ConvexPolygon& ConvexPolygon::operator= (const ConvexPolygon& T){
    if (this == &T) return *this;
    hull = T.Hull();
    return *this;
}
ConvexPolygon& ConvexPolygon::operator= (const std::vector<Point>& T){
    hull = convexHull(T);
    return *this;
}
ConvexPolygon::ConvexPolygon (const std::vector<Point>& T){
    hull = convexHull(T);
}
ConvexPolygon::ConvexPolygon (){
    hull = {};
}
void ConvexPolygon::setcol (const std::vector<double> new_color){
    color[0] = new_color[0]; color[1] = new_color[1]; color[2] = new_color[2];
}
std::vector<double> ConvexPolygon::getcol () const{
    return color;
}
// Useful information
// Cost: O(1)
int ConvexPolygon::num_vert () const{
    return hull.size();
}
// Cost: O(1)
int ConvexPolygon::num_edges () const{
    if (num_vert() >= 3) return num_vert();
    if (num_vert() == 2) return 1;
    return 0;
}
// Returns the area of a triangle
static double triangle_area(const Point& A, const Point& B, const Point& C){
    return abs(((B - A) ^ (C - A))/2);
}
// Returns the area of this polygon
double ConvexPolygon::area () const{
    double area = 0;
    for (int i = 1; i < num_vert()-1; i++){
        area += triangle_area(hull[0], hull[i], hull[i+1]);
    }
    return area;
}
double ConvexPolygon::perimeter () const{
    // Special cases
    if (num_vert() < 2) return 0;
    if (num_vert() == 2) return hull[0].distance(hull[1]);
    // Add each side
    double per = 0;
    for (int i = 0; i < hull.size()-1; i++) per += hull[i].distance(hull[i+1]);
    per += hull[hull.size()-1].distance(hull[0]);
    return per;
}
// Tells if all sides are equal, with a margin of 0.01%
static bool sideRegular (const std::vector<Point>& hull){
    double last_side = hull[0].distance(hull[1]);  
    double this_side;
    for (int i = 1; i < hull.size()-1; i++){
        this_side = hull[i].distance(hull[i+1]);  
        if (abs(last_side - this_side) > 0.0001*(last_side*this_side/2)) return false;
        last_side = this_side;
    }
    this_side = hull[hull.size()-1].distance(hull[0]);
    if (abs(last_side - this_side) > 0.0001*(last_side*this_side/2)) return false;
    // All tests succeeded
    return true;
}
// Tells if all angles are equal, with a margin of 1%
static bool angleRegular (const std::vector<Point>& hull){
    // Since it is convex, angles are between 0 and pi
    // In this range the cosine function is inyective
    // Therefore we can identify the angle with the cosine 
    // And since the sides are equal we can identify cosines with dot products
    double last_angle = (hull[2] - hull[1]) * (hull[1] - hull[0]);  
    double this_angle;
    for (int i = 1; i < hull.size()-2; i++){
        this_angle = (hull[i+2] - hull[i+1]) * (hull[i+1] - hull[i]);  
        if (abs(last_angle - this_angle) > 0.01*(last_angle*this_angle/2)) return false;
        last_angle = this_angle;
    }
    int i = hull.size() - 1;
    // Last two angles
    this_angle = (hull[0] - hull[i]) * (hull[i] - hull[i-1]);
    if (abs(last_angle - this_angle) > 0.01*(last_angle*this_angle/2)) return false;
    last_angle = this_angle;
    this_angle = (hull[1] - hull[0]) * (hull[0] - hull[i]);
    if (abs(last_angle - this_angle) > 0.01*(last_angle*this_angle/2)) return false;
    // All tests succedded
    return true;
}
// Tells if this polygon is regular
bool ConvexPolygon::isRegular () const{
    // Special cases
    if (num_vert() < 3) return true;
    return sideRegular(hull) and angleRegular(hull);
}
// Returns the centroid as a point
Point ConvexPolygon::centroid () const{
    Point G(0,0);
    for (Point p : hull) G += p;
    double n = hull.size();
    G *= 1/n;
    return G;
}
// Returns the set of all vertices
std::vector<Point> ConvexPolygon::Hull () const{
    return hull;
}

// Beginning of min / max functions
int ConvexPolygon::min_x (const std::vector<Point>& P) const{
    int minimum = 0;
    for (int i = 1; i < P.size(); i++){
        if (P[i].X() < P[minimum].X()) minimum = i;
    }
    return minimum;
}
int ConvexPolygon::min_y (const std::vector<Point>& P) const{
    int minimum = 0;
    for (int i = 1; i < P.size(); i++){
        if (P[i].Y() < P[minimum].Y()) minimum = i;
    }
    return minimum;
}
int ConvexPolygon::max_x (const std::vector<Point>& P) const{
    int maximum = 0;
    for (int i = 1; i < P.size(); i++){
        if (P[i].X() > P[maximum].X()) maximum = i;
    }
    return maximum;
}
int ConvexPolygon::max_y (const std::vector<Point>& P) const{
    int maximum = 0;
    for (int i = 1; i < P.size(); i++){
        if (P[i].Y() > P[maximum].Y()) maximum = i;
    }
    return maximum;
}
double ConvexPolygon::min_x_value (const std::vector<Point>& P) const{
    if (P.size() == 0) return 0;
    int minimum = 0;
    for (int i = 1; i < P.size(); i++){
        if (P[i].X() < P[minimum].X()) minimum = i;
    }
    return P[minimum].X();
}
double ConvexPolygon::min_y_value (const std::vector<Point>& P) const{
    if (P.size() == 0) return 0;
    int minimum = 0;
    for (int i = 1; i < P.size(); i++){
        if (P[i].Y() < P[minimum].Y()) minimum = i;
    }
    return P[minimum].Y();
}
double ConvexPolygon::max_x_value (const std::vector<Point>& P) const{
    if (P.size() == 0) return 0;
    int maximum = 0;
    for (int i = 1; i < P.size(); i++){
        if (P[i].X() > P[maximum].X()) maximum = i;
    }
    return P[maximum].X();
}
double ConvexPolygon::max_y_value (const std::vector<Point>& P) const{
    if (P.size() == 0) return 0;
    int maximum = 0;
    for (int i = 1; i < P.size(); i++){
        if (P[i].Y() > P[maximum].Y()) maximum = i;
    }
    return P[maximum].Y();
}
// End of min / max functions

/**

    Important Algorithms

*/

static bool angle_sorting (const Point& P1, const Point& P2){
    return P1.slope() < P2.slope();
}
// Changes the frame of reference of the points making the point Origin the new origin
static void move(std::vector<Point>& Polygon, const Point& Origin){
    for (Point& P : Polygon) P -= Origin; 
}
/**static void print(const std::vector<Point>& T){
    for (Point p : T) std::cout << p.X() << "," << p.Y() << " ";
    std::cout << std::endl;
}*/

// Returns a simple polygon (no intersections) for a given set of points
std::vector<Point> ConvexPolygon::simple_polygon (const std::vector<Point>& T) const{
    // Special cases
    if (T.size() < 2) return T;
    // Initialize
    std::vector<Point> Copy = T;
    // Find starting point
    int min_X = min_x(Copy);
    std::swap(Copy[0], Copy[min_X]);
    // Sort by the angles
    Point aux = Copy[0];
    move(Copy, aux);
    std::sort(++Copy.begin(), Copy.end(), angle_sorting);
    move(Copy, -aux);

    return Copy;
}

std::vector<Point> ConvexPolygon::graham_scan (const std::vector<Point>& T) const{
    // Special cases
    if (T.size() < 4) return T;
    // Initialize
    std::vector<Point> convexHull = {T[0]};
    // Traverse all points
    for (int i = 1; i < T.size(); i++){
        int ii = convexHull.size()-1; // Index of last point added
        // Remove the internal points 
        while (convexHull.size() > 1 and 
            not(Point().clockwise(convexHull[ii]-convexHull[ii-1], // Side 1
            T[i]-convexHull[ii]))) { // Side 2
                convexHull.pop_back();
                ii = convexHull.size()-1;
            }
        // Add the next point
        convexHull.push_back(T[i]);
    }
    return convexHull;
}

std::vector<Point> ConvexPolygon::convexHull (const std::vector<Point>& T) const{
    return graham_scan(simple_polygon(T));
}
static void append(std::vector<Point>& T1, const std::vector<Point>& T2){
    for (Point P : T2) T1.push_back(P);
}
// Returns segment that intersects AB if it exists
// Invariant: This segment is for sure between T[i..j]
static std::vector<Point> cut_points(const Point& A, const Point& B, const ConvexPolygon& T, int i, int j){
    return {A, B};
}
// Modifies V to contain the points that are inside the segment AB, both inclusive
static void check_inside(std::vector<Point>& V, const Point& A, const Point& B){

}
// Returns the cut points of the segment AB and T, if A or B are inside T they are returned as well
static std::vector<Point> cut_points(const Point& A, const Point& B, const ConvexPolygon& T){
    int m = (T.num_vert()-1)/2;
    std::vector<Point> result = cut_points(A, B, T, 0, m);
    append(result, cut_points(A, B, T, m, T.num_vert()-1));
    check_inside(result, A, B);
    return result;
}
ConvexPolygon ConvexPolygon::operator* (const ConvexPolygon& T) const{return ConvexPolygon();
    std::vector<Point> result;
    for (int i = 0; i < hull.size()-1; i++){
        append(result, cut_points(hull[i], hull[i+1], T));
    }
    append(result, cut_points(hull[hull.size()-1], hull[0], T));
    return ConvexPolygon(result);
}
ConvexPolygon ConvexPolygon::operator+ (const ConvexPolygon& T) const{
    std::vector<Point> Copy = T.Hull();
    append(Copy, hull);
    // Just create a convex polygon with all the points
    return ConvexPolygon(Copy);
}
static std::vector<Point> to_points(const std::vector<ConvexPolygon>& V){
    std::vector<Point> result;
    for (ConvexPolygon C : V) append(result, C.Hull());
    return result;
}
ConvexPolygon ConvexPolygon::bbox (const std::vector<ConvexPolygon>& V) const{
    std::vector<Point> points = to_points(V);
    double x_min = min_x_value(points);
    double y_min = min_y_value(points);
    double x_max = max_x_value(points);
    double y_max = max_y_value(points);

    Point LL(x_min, y_min);
    Point LR(x_max, y_min); 
    Point UR(x_max, y_max);
    Point UL(x_min, y_max);
    return ConvexPolygon({LL, LR, UR, UL});
}
bool ConvexPolygon::isInside (const ConvexPolygon& T) const{
    std::vector<Point> T_hull = T.Hull();
    for (Point P : hull) if (not isInside(T_hull, P)) return false;
    return true;
}
// Tells if a point is inside the triangle T[0] T[i1] T[i2]
static bool insd3(const std::vector<Point>& T, int i1, int i2, const Point& P){
    bool left1 = (T[i1] - T[0]).clockwise(P - T[i1]);
    bool left2 = (T[i2] - T[i1]).clockwise(P - T[i2]);
    bool left3 = (T[0] - T[i2]).clockwise(P - T[0]);
    return left1 and left2 and left3;
}
/** Tells if P is inside the portion delimited by the following polygon:
    - Start in T[0]
    - The next vertex is T[i1]
    - Then we have all the vertices T[i1..i2]
    - And finally the side T[i2]T[0]
*/
static bool isInterior (const std::vector<Point>& T, int i1, int i2, const Point& P){
    // Base Case
    if (i1 == i2 - 1) return insd3(T, i1, i2, P);
    // Divide
    int mid = (i2+i1)/2;
    Point side = T[mid] - T[0];
    Point point = P - T[mid];
    // Conquer
    if (not side.clockwise(point)) return isInterior(T, i1, mid, P); // To the right
    return isInterior(T, mid, i2, P); // To the left
}
bool ConvexPolygon::isInside (const std::vector<Point>& T, const Point& P) const{
    if (T.size() < 3) return false;
    return isInterior(T,1,T.size()-1,P);
}
