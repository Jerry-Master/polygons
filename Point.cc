#include "Point.h"

// Constructor
Point& Point::operator= (const Point& P){
    x = P.X(); y = P.Y();
    return *this;
}
Point::Point (double x2, double y2) : x(x2), y(y2) {}

// Coordinates
double Point::X () const {
    return x;
}
double Point::Y () const {
    return y;
}

// Important values
// Quadrance, just to avoid square roots
double Point::quadrance() const {
    return x*x + y*y;
} 
// To the x-axis
double Point::slope() const {
    if (abs(x) < EPSILON) return INFINITY;
    return y/x;
}
// (first goes the x and then the y) 1: ++, 2: -+, 3: --, 4: +-, 5: axis
int Point::quadrant() const {
    if (x > 0 and y > 0) return 1;
    if (x < 0 and y > 0) return 2;
    if (x < 0 and y < 0) return 3;
    if (x > 0 and y < 0) return 4;
    // Dealing with precision issues around the axis
    if (abs(x) < EPSILON or abs(y) > EPSILON) return 5;
    else assert(false);
} 
double Point::distance(const Point& P) const{
    return sqrt((P - *this).quadrance());
}

// Add, substract, cross-product, dot product, clockwise orientation
Point& Point::operator+= (const Point& P){
    x += P.X(); y += P.Y();
    return *this;
}
Point Point::operator+ (const Point& P) const{
    Point _aux = *this;
    _aux += P;
    return _aux;
}
Point& Point::operator-= (const Point& P){
    x -= P.X(); y -= P.Y();
    return *this;
}
Point Point::operator- (const Point& P) const{
    Point _aux = *this;
    _aux -= P;
    return _aux;
}
Point Point::operator- () const{
    Point _aux(-x, -y);
    return _aux;
}
// Multiplication by a scalar
Point& Point::operator*= (double lambda){
    x *= lambda; y *= lambda;
    return *this;
}
// Cross-product
double Point::operator^ (const Point& P) const{
    return x * P.Y() - y * P.X();
}
// Dot-product
double Point::operator* (const Point& P) const{
    return x * P.X() + y * P.Y();
}
bool Point::operator== (const Point& P) const{
    return x == P.X() and y == P.Y();
}
bool Point::clockwise (const Point& P1, const Point& P2) const{
    return (P1 ^ P2) >= 0;
}
bool Point::clockwise (const Point& P) const{
    return clockwise(*this, P) > 0;
}
