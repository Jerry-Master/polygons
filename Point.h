#ifndef _POINT
#define _POINT
#define EPSILON 1e-12 

#include <cmath>
#include <cassert>

/**
This class provides all the useful functions to work with points. 
We can also identify them as vectors taking the origin as reference,
therefore all the useful operations with vectors are also implemented.

We can consider the cost of its functions as O(1)
*/

class Point{
    public:
        // Constructor
        Point& operator= (const Point& P);
        Point (double x = 0, double y = 0);

        // Coordinates
        double X () const;
        double Y () const;
        
        // Important values
        double quadrance() const; // Quadrance, just to avoid square roots
        double slope() const; // To the x-axis
        // (first goes the x and then the y) 1: ++, 2: -+, 3: --, 4: +-, 5: axis
        // To distinguish the different slopes
        int quadrant() const; 
        // Distance from this to P
        double distance(const Point& P) const;

        // Add, substract, cross-product, dot product, clockwise orientation
        Point& operator+= (const Point& P);
        Point operator+ (const Point& P) const;
        Point& operator-= (const Point& P);
        Point operator- (const Point& P) const;
        Point operator- () const;
        Point& operator*= (double lambda); // Multiplication by a scalar
        double operator^ (const Point& P) const; // Cross
        double operator* (const Point& P) const; // Dot
        bool operator== (const Point& P) const;
        // Taking the origin as reference
        // True: P1^P2 > 0
        // False: P1^P2 < 0
        // Prec: Neither P1 nor P2 are the origin
        bool clockwise (const Point& P1, const Point& P2) const;
        bool clockwise (const Point& P) const;

    private:
        double x, y;
};

#endif
