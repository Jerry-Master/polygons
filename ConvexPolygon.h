#ifndef _CONV
#define _CONV

#include "Point.h"
#include <vector>
#include <algorithm>

/**
To clarify things, in the followings n = hull.size() and m = T.size()
Or if there are two polygons T1 and T2, n = T1.size() and m = T2.size()
Or n is simply the size of the attribute to the function

The vertices are ordered counterclockwise
*/

class ConvexPolygon{
    public:
        // Constructor
        ConvexPolygon& operator= (const ConvexPolygon& T); // Check
        ConvexPolygon& operator= (const std::vector<Point>& T); // Check
        ConvexPolygon (const std::vector<Point>& T); // Check
        ConvexPolygon (); // Check
        // Changes the color
        void setcol (const std::vector<double> new_color);
        std::vector<double> getcol () const;

        /**
            Both the intersection and the union doesn't change the father polygon,
            but they return a reference to the polygon created in order to be able 
            to concatenate operations
        */

        // Intersection 
        // Cost: O(n^2)
        ConvexPolygon operator* (const ConvexPolygon& T) const;
        // Union
        // Cost: O((m+n)log(m+n))
        ConvexPolygon operator+ (const ConvexPolygon& T) const; // Check

        // Smallest Bounding Box that contains the elements
        // Cost: O(n), n = total number of vertices
        ConvexPolygon bbox (const std::vector<ConvexPolygon>& V) const; // Check?
        // Tells if this polygon is inside the polygon T
        // Cost: O(n log(m)) 
        bool isInside (const ConvexPolygon& T) const; // Check

        // Useful information
        // Cost: O(1)
        int num_vert () const; // Check
        // Cost: O(1)
        int num_edges () const; // Check
        // Cost: O(n) flops
        double area () const; // Check
        // Cost: O(n) flops
        double perimeter () const;
        // Considering the emptygon, monogon and dygon as regular
        // Cost: O(n) flops
        bool isRegular () const; // Check
        // Cost: O(n) additions
        Point centroid () const; // Check
        // Returns the set of all vertices
        std::vector<Point> Hull () const; // Check

    private:
        /** 
        Internal representation:
        hull[i] is joined to hull[i+1] y hull[i-1]
        We can consider hull.size() == 0 and -1 == hull.size() - 1
        And hull[0] has the lowest x-coordinate
        */
        std::vector<Point> hull;
        std::vector<double> color = {0.0, 0.0, 0.0}; // Default is black

        //Useful methods either for constructor or other public methods
        
        // Returns the index/coordinate of the point with the minimum/maximum x/y-coordinate 
        // Cost: O(n)
        int min_x (const std::vector<Point>& P) const; // Check
        int min_y (const std::vector<Point>& P) const; // Check
        int max_x (const std::vector<Point>& P) const; // Check
        int max_y (const std::vector<Point>& P) const; // Check
        // For the Bounding Box function
        double min_x_value (const std::vector<Point>& P) const; // Check
        double min_y_value (const std::vector<Point>& P) const; // Check  
        double max_x_value (const std::vector<Point>& P) const; // Check
        double max_y_value (const std::vector<Point>& P) const; // Check
        //Cost: O(n log(n))
        std::vector<Point> simple_polygon (const std::vector<Point>& T) const; // Check
        // Cost: O(n + m), for the union
        std::vector<Point> simple_polygon (const ConvexPolygon& T1, const ConvexPolygon& T2) const;
        // Cost: O(n)
        std::vector<Point> graham_scan (const std::vector<Point>& T) const; // Check
        // Cost: O(n log(n)), it is simple_polygon + graham_scan
        std::vector<Point> convexHull (const std::vector<Point>& T) const; // Check
        // Cost: O(log(n))
        bool isInside (const std::vector<Point>& T, const Point& P) const; // Check
        // Returns the points of this inside this, and the index of the beginning and ending in i and j
        std::vector<Point> pointsInside(const std::vector<Point>& T, int i, int j) const;
};

#endif
