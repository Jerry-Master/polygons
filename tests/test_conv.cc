#include "../Point.h"
#include "../ConvexPolygon.h"

#include <iostream>
#include <map>
#include <sstream>
using namespace std;

map<string, ConvexPolygon> pols;

void add_polygon(){
    string name;
    cin >> name;
    vector<Point> v;
    string line;
    getline(cin, line);
    istringstream iss(line);
    double x, y;
    while (iss >> x >> y) v.push_back(Point(x,y));
    pols[name] = v;
}

void remove(){
    string name;
    cin >> name;
    pols.erase(name);
}
// Assigns the first to the second
void assign(){
    string name1, name2;
    cin >> name1 >> name2;
    pols[name2] = pols[name1];
}
// Shows all the vertices of T in order, in the form (x1,y1) (x2,y2)  
void print(const ConvexPolygon& T){
    vector<Point> v = T.Hull();
    for (Point p : v) cout << '(' << p.X() << ',' << p.Y() << ") ";
    cout << endl;
}

void Union(){
    string name1, name2;
    cin >> name1 >> name2;
    print(pols[name1] + pols[name2]);
}

void intersection(){
    string name1, name2;
    cin >> name1 >> name2;
    print(pols[name1] * pols[name2]);
}

void bbox(){
    ConvexPolygon box;
    string name;
    while (cin >> name) box = ConvexPolygon().bbox({box, pols[name]});
    print(box);
}
// Tells if the first is inside the second
void isInside(){
    string name1, name2;
    cin >> name1 >> name2;
    bool insd = pols[name1].isInside(pols[name2]);
    if (insd) cout << "yes" << endl;
    else cout << "no" << endl;
}

void num_vert(){
    string name;
    cin >> name;
    cout << pols[name].num_vert() << endl;
}

void num_edges(){
    string name;
    cin >> name;
    cout << pols[name].num_edges() << endl;
}

void area(){
    string name;
    cin >> name;
    cout << pols[name].area() << endl;
}

void isRegular(){
    string name;
    cin >> name;
    cout << (pols[name].isRegular() ? "yes" : "no") << endl;
}

void centroid(){
    string name;
    cin >> name;
    Point centroid = pols[name].centroid();
    cout << '(' <<centroid.X() << ',' << centroid.Y() << ')' << endl;
}

void Hull(){
    string name;
    cin >> name;
    print(pols[name]);
}

int main(){
    cout.setf(ios::fixed);
    cout.precision(2);

    string inst;
    while (cin >> inst){
        if (inst == "polygon") add_polygon();
        else if (inst == "remove") remove();
        else if (inst == "assign") assign();

        else if (inst == "+") Union();
        else if (inst == "*") intersection();

        else if (inst == "bbox") bbox();
        else if (inst == "isInside") isInside();

        else if (inst == "num_vert") num_vert();
        else if (inst == "num_edges") num_edges();
        else if (inst == "area") area();
        else if (inst == "isRegular") isRegular();
        else if (inst == "centroid") centroid();
        else if (inst == "Hull") Hull();
    }
}
