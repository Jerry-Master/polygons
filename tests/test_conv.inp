polygon p1 -2.06 2.23   -3.74 1.67   -2.16 -2.59   1.32 1.49   -1.54 1.07   -2.28 0.63   -0.82 0.63   -1 -1   -1.52 -0.59   -1 2   0.6 0.43   -2.96 -0.23   -4.54 -1.17   0.72 -1.79   -0.5 2.63   -2.8 3.35
Hull p1
num_vert p1
num_edges p1
area p1
isRegular p1
centroid p1
polygon p2 -2.77 -1.26
polygon p3 -4.36 -3.46
isInside p2 p1
isInside p3 p1
polygon p4 -0.9385 -2.5511   -0.0511 -4.5283   2.1049 -4.7483   3.3735 -2.9912   2.486 -1.014   0.33 -0.794
isRegular p4
* p1 p4
+ p1 p4
area p4