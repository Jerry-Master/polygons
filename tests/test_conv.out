(-4.54,-1.17) (-2.16,-2.59) (0.72,-1.79) (1.32,1.49) (-0.50,2.63) (-2.80,3.35) (-3.74,1.67) 
7
7
24.06
no
(-1.67,0.51)
yes
no
yes

(-4.54,-1.17) (-0.05,-4.53) (2.10,-4.75) (3.37,-2.99) (2.49,-1.01) (1.32,1.49) (-0.50,2.63) (-2.80,3.35) (-3.74,1.67) 
12.20
